File: app/api/v1/exchange_rates.py
"""
Exchange Rate API endpoints
Enhanced with history tracking and chart data
"""
import httpx
from datetime import datetime, date, timedelta
from typing import Optional
from fastapi import APIRouter, HTTPException, Depends, Query
from sqlalchemy.orm import Session
from sqlalchemy import desc

from app.database.config import get_db
from app.models.exchange_rate import ExchangeRateHistory
from app.utils.logger import setup_logger

logger = setup_logger(__name__)

router = APIRouter()


@router.get("/current")
async def get_current_rates(db: Session = Depends(get_db)):
    """
    Get current USD/KRW and CNY/KRW exchange rates.
    Also saves to history if not already saved today.
    """
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            response = await client.get(
                "https://open.er-api.com/v6/latest/USD"
            )
            if response.status_code != 200:
                raise HTTPException(status_code=502, detail="환율 API 호출 실패")

            data = response.json()
            rates = data.get("rates", {})

            krw_per_usd = rates.get("KRW", 0)
            cny_per_usd = rates.get("CNY", 0)

            # Calculate CNY/KRW
            krw_per_cny = round(krw_per_usd / cny_per_usd, 2) if cny_per_usd > 0 else 0

            # Save to history (once per day)
            today = date.today()
            existing = (
                db.query(ExchangeRateHistory)
                .filter(
                    ExchangeRateHistory.rate_date == today,
                    ExchangeRateHistory.currency_pair == "USD_KRW",
                )
                .first()
            )

            if not existing and krw_per_usd > 0:
                for pair, rate_val in [("USD_KRW", krw_per_usd), ("CNY_KRW", krw_per_cny)]:
                    entry = ExchangeRateHistory(
                        currency_pair=pair,
                        rate=round(rate_val, 2),
                        rate_date=today,
                    )
                    db.add(entry)
                db.commit()
                logger.info(f"환율 히스토리 저장: USD/KRW={krw_per_usd}, CNY/KRW={krw_per_cny}")

            return {
                "status": "success",
                "data": {
                    "USD_KRW": round(krw_per_usd, 2),
                    "CNY_KRW": krw_per_cny,
                    "USD_CNY": round(cny_per_usd, 4),
                    "updated_at": data.get("time_last_update_utc", datetime.now().isoformat()),
                },
            }
    except httpx.TimeoutException:
        logger.warning("환율 API 타임아웃")
        # Fallback to latest from DB
        latest = _get_latest_from_db(db)
        if latest:
            return {"status": "success", "data": latest, "source": "cache"}
        return {
            "status": "error",
            "message": "환율 정보를 가져오는데 시간이 초과되었습니다.",
            "data": None,
        }
    except Exception as e:
        logger.error(f"환율 조회 실패: {e}")
        latest = _get_latest_from_db(db)
        if latest:
            return {"status": "success", "data": latest, "source": "cache"}
        return {
            "status": "error",
            "message": "환율 정보를 가져올 수 없습니다.",
            "data": None,
        }


@router.get("/history")
async def get_rate_history(
    currency_pair: str = Query("USD_KRW", regex="^(USD_KRW|CNY_KRW)$"),
    days: int = Query(30, ge=7, le=365),
    db: Session = Depends(get_db),
):
    """
    Get exchange rate history for chart display.
    Returns daily rates for the specified period.
    """
    start_date = date.today() - timedelta(days=days)

    history = (
        db.query(ExchangeRateHistory)
        .filter(
            ExchangeRateHistory.currency_pair == currency_pair,
            ExchangeRateHistory.rate_date >= start_date,
        )
        .order_by(ExchangeRateHistory.rate_date)
        .all()
    )

    return {
        "status": "success",
        "currency_pair": currency_pair,
        "days": days,
        "data": [
            {
                "date": h.rate_date.isoformat(),
                "rate": h.rate,
            }
            for h in history
        ],
    }


@router.post("/fetch-history")
async def fetch_and_store_history(
    days: int = Query(30, ge=1, le=90),
    db: Session = Depends(get_db),
):
    """
    Fetch historical rates and store in DB.
    Uses the free API to backfill data.
    """
    saved_count = 0
    errors = 0

    try:
        async with httpx.AsyncClient(timeout=15.0) as client:
            for i in range(days):
                target_date = date.today() - timedelta(days=i)

                # Skip if already have data
                existing = (
                    db.query(ExchangeRateHistory)
                    .filter(
                        ExchangeRateHistory.rate_date == target_date,
                        ExchangeRateHistory.currency_pair == "USD_KRW",
                    )
                    .first()
                )
                if existing:
                    continue

                try:
                    url = f"https://open.er-api.com/v6/latest/USD"
                    response = await client.get(url)
                    if response.status_code != 200:
                        errors += 1
                        continue

                    data = response.json()
                    rates = data.get("rates", {})
                    krw = rates.get("KRW", 0)
                    cny = rates.get("CNY", 0)

                    if krw > 0 and cny > 0:
                        krw_per_cny = round(krw / cny, 2)

                        for pair, rate_val in [("USD_KRW", round(krw, 2)), ("CNY_KRW", krw_per_cny)]:
                            entry = ExchangeRateHistory(
                                currency_pair=pair,
                                rate=rate_val,
                                rate_date=target_date,
                            )
                            db.add(entry)
                        saved_count += 1

                except Exception:
                    errors += 1
                    continue

        db.commit()
        logger.info(f"환율 히스토리 백필 완료: {saved_count}일 저장, {errors}건 오류")

        return {
            "status": "success",
            "message": f"환율 히스토리 {saved_count}일치 저장 완료",
            "data": {"saved": saved_count, "errors": errors},
        }

    except Exception as e:
        logger.error(f"환율 히스토리 백필 실패: {e}")
        raise HTTPException(status_code=500, detail=f"히스토리 백필 실패: {str(e)}")


def _get_latest_from_db(db: Session) -> Optional[dict]:
    """Get latest rates from DB as fallback"""
    usd = (
        db.query(ExchangeRateHistory)
        .filter(ExchangeRateHistory.currency_pair == "USD_KRW")
        .order_by(desc(ExchangeRateHistory.rate_date))
        .first()
    )
    cny = (
        db.query(ExchangeRateHistory)
        .filter(ExchangeRateHistory.currency_pair == "CNY_KRW")
        .order_by(desc(ExchangeRateHistory.rate_date))
        .first()
    )

    if usd:
        return {
            "USD_KRW": usd.rate,
            "CNY_KRW": cny.rate if cny else 0,
            "USD_CNY": 0,
            "updated_at": usd.rate_date.isoformat(),
        }
    return None

File: workers-api/src/routes/exchange-rates.ts
/**
 * Exchange Rate Routes - /api/v1/exchange-rates
 */
import { Hono } from "hono";
import { drizzle } from "drizzle-orm/d1";
import { eq, desc, and, gte } from "drizzle-orm";
import { exchangeRateHistory } from "../db/schema";
import { authMiddleware } from "../middleware/auth";
import type { Env } from "../types";

const rates = new Hono<{ Bindings: Env }>();

/**
 * GET /exchange-rates/current - 현재 환율 (인증 불필요 - 대시보드 위젯용)
 */
rates.get("/current", async (c) => {
  const db = drizzle(c.env.DB);

  // KV 캐시 확인
  if (c.env.CACHE) {
    try {
      const cached = await c.env.CACHE.get("exchange_rates_current", "json");
      if (cached) {
        return c.json({ status: "success", data: cached });
      }
    } catch { /* KV not available */ }
  }

  // 네이버 금융 API에서 실시간 환율 가져오기
  try {
    const [usdRes, cnyRes] = await Promise.all([
      fetch("https://api.stock.naver.com/marketindex/exchange/FX_USDKRW"),
      fetch("https://api.stock.naver.com/marketindex/exchange/FX_CNYKRW"),
    ]);

    const usdData = (await usdRes.json()) as any;
    const cnyData = (await cnyRes.json()) as any;

    if (usdData.exchangeInfo && cnyData.exchangeInfo) {
      const usdKrw = parseFloat(usdData.exchangeInfo.closePrice.replace(/,/g, ""));
      const cnyKrw = parseFloat(cnyData.exchangeInfo.closePrice.replace(/,/g, ""));
      const usdCny = Math.round((usdKrw / cnyKrw) * 100) / 100;
      const today = new Date().toISOString().split("T")[0];

      const result = {
        USD_KRW: usdKrw,
        CNY_KRW: cnyKrw,
        USD_CNY: usdCny,
        updated_at: usdData.exchangeInfo.localTradedAt || new Date().toISOString(),
        source: "naver",
      };

      // D1에 저장
      await db
        .insert(exchangeRateHistory)
        .values([
          { currencyPair: "USD_KRW", rate: usdKrw, rateDate: today, source: "naver" },
          { currencyPair: "CNY_KRW", rate: cnyKrw, rateDate: today, source: "naver" },
        ])
        .onConflictDoNothing();

      // KV에 1시간 캐시
      if (c.env.CACHE) {
        try {
          await c.env.CACHE.put("exchange_rates_current", JSON.stringify(result), {
            expirationTtl: 3600,
          });
        } catch { /* KV not available */ }
      }

      return c.json({ status: "success", data: result });
    }
  } catch (e) {
    // 네이버 API 실패 시 DB에서 최신 데이터
  }

  // 폴백: DB에서 최신 환율
  const usd = await db
    .select()
    .from(exchangeRateHistory)
    .where(eq(exchangeRateHistory.currencyPair, "USD_KRW"))
    .orderBy(desc(exchangeRateHistory.rateDate))
    .limit(1);

  const cny = await db
    .select()
    .from(exchangeRateHistory)
    .where(eq(exchangeRateHistory.currencyPair, "CNY_KRW"))
    .orderBy(desc(exchangeRateHistory.rateDate))
    .limit(1);

  return c.json({
    status: "success",
    data: {
      USD_KRW: usd[0]?.rate || 0,
      CNY_KRW: cny[0]?.rate || 0,
      USD_CNY: 0,
      updated_at: usd[0]?.createdAt || null,
    },
  });
});

/**
 * GET /exchange-rates/history - 환율 히스토리 (차트용)
 */
rates.get("/history", authMiddleware, async (c) => {
  const pair = c.req.query("pair") || "USD_KRW";
  const days = parseInt(c.req.query("days") || "30");

  const db = drizzle(c.env.DB);
  const startDate = new Date();
  startDate.setDate(startDate.getDate() - days);
  const startDateStr = startDate.toISOString().split("T")[0];

  const history = await db
    .select()
    .from(exchangeRateHistory)
    .where(
      and(
        eq(exchangeRateHistory.currencyPair, pair),
        gte(exchangeRateHistory.rateDate, startDateStr)
      )
    )
    .orderBy(exchangeRateHistory.rateDate);

  return c.json({ status: "success", data: history });
});

export default rates;
